tik/lib/CartographerKernel.cpp

```diff
@@ -291,500 +281,926 @@ namespace TraceAtlas::tik
-                            break;
+                            // may belong to a subkernel, which is not an export
+                            if (embeddedKernels.find(useInst->getParent()->getParent()) == embeddedKernels.end())
+                            {
+                                auto sExtVal = GetValueID(inst);
+                                if (kernelIE.find(sExtVal) == kernelIE.end())
+                                {
+                                    KernelExports.push_back(sExtVal);
+                                    kernelIE.insert(sExtVal);
+                                }
+                                else if (find(KernelImports.begin(), KernelImports.end(), sExtVal) != KernelImports.end())
+                                {
+                                    throw AtlasException("Import needs to be an export!");
+                                }
+                            }
                         }
                     }
-                    else
+                }
+                // now we have to check the block successors
+                // if this block can exit the kernel, that means we are replacing a block in the source bitcode that may be left with no predecessors
+                // but there may still be users of its values. So they need to be exported
+                for (auto succ : successors(block))
+                {
+                    if (scopedBlocks.find(succ) == scopedBlocks.end())
                     {
-                        throw AtlasException("Non-instruction user detected");
+                        // this block can exit
+                        // if the value uses extend beyond this block, export it
+                        for (auto use : inst->users())
+                        {
+                            if (auto outInst = dyn_cast<Instruction>(use))
+                            {
+                                if (outInst->getParent() != inst->getParent()) // && scopedBlocks.find(inst->getParent()) == scopedBlocks.end() )
+                                {
+                                    auto sExtVal = GetValueID(inst);
+                                    if (kernelIE.find(sExtVal) == kernelIE.end())
+                                    {
+                                        KernelExports.push_back(sExtVal);
+                                        kernelIE.insert(sExtVal);
+                                    }
+                                    else if (find(KernelImports.begin(), KernelImports.end(), sExtVal) != KernelImports.end())
+                                    {
+                                        throw AtlasException("Import needs to be an export!");
+                                    }
+                                }
+                            }
+                        }
                     }
                 }
             }
         }
-        if (Entrances.empty())
```

Comment:
This check is still very important until kernels are parsed correctly 100% of the time.

Reply:
This check has been moved to the top of ConstructFunctionSignature()